### ISSUE TITLE
Define booking state (NEW, CHANGED, UPDATEREQUESTED etc.)

### ISSUE THREAD

**Issue by efel85** [efel85/TOMP-API#9]
submitted by Pim van Toolen

**Comment by pimmeh** [comment]
It is stated that this should be the life-cycle of the booking. Given that, I suggest the the followin states and rationales as a minimum:

- **OPTION**
When a user wants to make a trip plan and has started the booking process, under water the MaaS Provider may create a booking by the transport operator before the user finishes the booking process. This is usefull when there is a limited number of, for instance, bikes. This way, the bike is reserved for like 15 minutes in which the user has the time to finish the booking process. Bookings in this state will automatically put in the state CANCELLED after x minutes if the state is not adjusted to BOOKED within that time frame.
- **BOOKED**
The trip is booked with the transport operator.
- **CANCELLED**
After the booking has been cancelled (for whatever reaseon by whatever party) or when it is in the state OPTION for longer than it is allowed, the booking is CANCELLED and cannot be changed by the user or MaaS-provider anymore. A new booking is needed. All the reserved transports (bikes, rides, seats, cars, etcetera) are freed so others can book them.
- **STARTED**
The booking has started, it is assumed that the user is travelling with the transport operator as per the booking. 
- **FINISHED**
This state can only be reached after the STARTED state. It is assumed the user actually travelled with this transport operator and his/her leg has succesfully ended. This is the final state of the booking and can no longer be updated when in this state.

There were also NEW, PAID and UPDATES states in the current version that I removed.
- NEW is, I think, pointless as a booking should then immediatly be put in a next state, such as BOOKED or OPTION.
- PAID should be a different field, if wanted, but not be part of the booking state. Because a booking can go an entire life cycle of states and at different times, a payment can be made / wanted. Also, is this payed from the perspective of the TO or MP? I suggest a new issue if we want to discuss this.
- UPDATED is also removed, I see added functionality. It might be possible to include a new field { "last_updated": "timestamp" } to a booking to check when it was last updated? I suggest a new issue if we want to discuss this.

**Comment by efel85** [comment]
We should add relations to different modalities to clarify the semantics of planning, booking, execution

**Comment by pimmeh** [comment]
After the last TOM-meeting, we decided that we should replace the OPTIONAL state with something like 'PENDING'. This is a temporary state where an asset is reserved for a very small amount of time so that the MaaS-provider can reserve all the assets in a complete journey, before giving them the final stated BOOKED. This prevents the scenario that the user clicks 'book entire journey' (for example, consisting out of 4 legs and transport operators), that the MaaS Provider books the first two legs and that in the meantime the third asset is reserved for a different user. Only after that all the legs are in the state PENDING will the MaaS provider book all the assets and change them to BOOKED. This results in the following states:

- **PENDING**
When a user wants to book a journey (consisting out of multiple legs) and has clicked 'book entire journey', under water the MaaS Provider creates all bookings at the relevant transport operators in the PENDING state. This is usefull when there is a limited number of, for instance, bikes. This way, the bike is reserved for like 10 seconds in which the messages have the time to travel from MaaS-provider to all the transport operators and receive a reply. When all the bookings are confirmed in the state PENDING, the MaaS-provider will change all bookings to the state BOOKED. Bookings in the PENDING state will automatically be deleted (not changed to CANCELLED) if the state is not adjusted to BOOKED within the designated time frame.
- **BOOKED**
The trip is booked with the transport operator.
- **CANCELLED**
After the booking has been cancelled (for whatever reaseon by whatever party), the booking is CANCELLED and cannot be changed by the user or MaaS-provider anymore. A new booking is needed. All the reserved transports (bikes, rides, seats, cars, etcetera) are freed so others can book them.
- **STARTED**
The booking has started, it is assumed that the user is travelling with the transport operator as per the booking. 
- **FINISHED**
This state can only be reached after the STARTED state. It is assumed the user actually travelled with this transport operator and his/her leg has succesfully ended. This is the final state of the booking and can no longer be updated when in this state.

**Comment by JRobertoReyesG** [comment]
- I suggest to use the term BOOKING only to refer to the process of making a RESERVATION. All the suggested states therefore would refer to the RESERVATION. I've changed the terminology in the process flows of the blueprint accordingly (subject to approval of course). All "BOOKING" calls would need to be changed (i.e. Booking {id}) 
- To avoid further confusion with the terms "book / booked / booking", I suggest to use CONFIRMED instead of BOOKED.
- STARTED state <-- is this state needed? As with the FINISHED state, it is assumed that the user actually travelled according to the reservation that was CONFIRMED. The only relevant point is the starting_time as stated in the RESERVATION_ID
- I agree with the other suggested states.

**Comment by JRobertoReyesG** [comment]
As promised in our last meeting (24-07-2019), I looked for similar terminology in other transportation standards. I found some references with what I think is relevant terminology for the API definition, but I couldn't find anything specific for "booking status". In the attachment you can find a document that I started with a list of terms and definitions including the booking lstatus listed in this issue. Also in the attachment, one glossary from RNE and another from IATA. In those I marked some more terms that I think could be applicable for our work. Please check the documents out and let's discuss their usability.
[List of definitions_2019-07-31.docx](https://github.com/efel85/TOMP-API/files/3452016/List.of.definitions_2019-07-31.docx)
[IATA-Passenger-Glossary-of-Terms.xlsx](https://github.com/efel85/TOMP-API/files/3451963/IATA-Passenger-Glossary-of-Terms.xlsx)
[RNE_NS_CID_Glossary_2019_final.xlsx](https://github.com/efel85/TOMP-API/files/3451964/RNE_NS_CID_Glossary_2019_final.xlsx)

**Comment by efel85** [comment]
![Booking States](https://user-images.githubusercontent.com/23665014/62549150-44afeb00-b868-11e9-9088-525bb3215e81.png)
Above result after discussions in two previous working groups, see also #41

**Comment by efel85** [comment]
Updated the booking state sequence diagram during the meeting on 07/08

![190807 Booking States-update](https://user-images.githubusercontent.com/23665014/62612280-5eecd600-b907-11e9-8f35-608d17fe9504.png)

**Comment by efel85** [comment]
**Booking states sequence diagramme description**

_Planning phase_
In the planning phase, a MSP can check the real-time availability of assets from a TO. In this way, a MSP can offer their Users an overview of which assets and options are currently available following the User's request (for a specific mode, a specific location or other User conditions). A time-to-live can optionally be added to the response to show the User how long the information will be valid for.

_Booking phase_
Once the User has narrowed down their selection, the MSP can send an availability request to the TO for a specific asset (or asset type) selection. This changes the Booking State to PENDING and temporarily 'freezes' an asset while the User is finalizing the selection (i.e. while the User is having to choose multiple options for multiple legs of a journey). A time-to-live in the availability confirmation response is mandatory.

If a User decides to go for other options than the one(s) which was or were narrowed down, the PENDING state can be cancelled by the MSP. The Booking State is changed to RELEASED.

If the expiry time for the PENDING state is reached (as defined in the time-to-live in the availability confirmation), because the User has not (yet) made a selection, the Booking State changes to EXPIRED and the corresponding asset(s) are no longer 'frozen' for the specific request. 

If a User confirms the selection of a given option, the asset (or asset type) is requested from the TO and the Booking State changes to CONFIRMED.

_Trip Execution phase_
Once the confirmed asset is in use, the Booking State is changed to STARTED.

If the asset confirmation is cancelled by the MSP (which could also happen upon request from the User), the Booking State changes to CANCELLED, and the corresponding terms and conditions for cancellations between TOs and MSPs apply.

If the asset confirmation is cancelled by the TO (in case of a broken down vehicle, late return etc.), the  Booking State changes to CANCELLED, and the corresponding terms and conditions for cancellations between TOs and MSPs apply.

Once the asset is returned, the leg is considered completed and the Booking State is changed to FINISHED.

**Comment by pimmeh** [comment]
Perfect!

**Comment by edwinvandenbelt** [comment]
Happy flow is clear. I see some non-happy paths. In the SD it's assumed that there is 2-way communication. What if something happens and a 'cancel' or 'expire' from the TO is not arriving the MSP? There should be a facility to respond 'Cancelled' or 'Expired' on requests from the MSP.

**Comment by pimmeh** [comment]
Please no, we have TCP/IP for that is messages are sent. Otherwise, we are going to duplicate that technology.

**Comment by edwinvandenbelt** [comment]
ok! The responses are posted on the webhooks. Clear.

**Comment by efel85** [comment]
Update with correct depiction of alternatives

![Booking States](https://user-images.githubusercontent.com/23665014/63949201-4f7e2a00-ca7a-11e9-9032-f87022c42eb8.png)

**Comment by efel85** [comment]
@edwinvandenbelt Edwin has offered to update the code this week following the sequence diagram above.

**Comment by JRobertoReyesG** [comment]
Update of the BOOKING operational flow with embedded booking states' transition flows.

![TO-MP API blocks visualization _ 2019_09_10_BOOKING](https://user-images.githubusercontent.com/51329434/64701525-39745e80-d4a9-11e9-855d-a760c714d4ab.jpg)

**Comment by JRobertoReyesG** [comment]
"BOOKING" refers to the process of making a RESERVATION (see my comment on 1 July), I changed the terminology in the process flow but the "booking" objects in the swagger file still need to be updated (and I noticed now that I also didn't update the foot-note on the process flow diagram, it has been corrected now). 

The state diagram above is a suggestion to attach an asset id to any RESERVATION, so any additions derived from it would also need to be updated, isn't it @edwinvandenbelt ?

**Comment by edwinvandenbelt** [comment]
Maybe 'reserve' isn't the correct word to use. Probably 'claim-asset' would be better. Purpose is to assign a specific asset to a booking. It should before the start of the trip, when there is not yet an asset assigned (this is in case of upfront booked travelling & booked a type of asset).

**Comment by edwinvandenbelt** [comment]
The STD without 'claim-asset': 
![image](https://user-images.githubusercontent.com/10400054/64861440-3c4d8b80-d630-11e9-90c2-9263e805d0aa.png)

can be edited using: https://cruise.eecs.uottawa.ca/umpleonline/#genArea
// UML state diagram for a Booking, represented in Umple

class Booking {
  state {
    availibility_checked { 
      availability-request -> pending;
    }
    pending {
        asset-request -> confirmed;
        cancel-request -> released;
        expired -> expired;
    }
    confirmed {
        cancel-booking -> cancelled;
        in-preparation -> preparing;
        in-use-request -> in_use;
    }
    preparing {
       in-use-confirm -> in_use;
    }
    in_use {
        start-pause -> paused;
        start-finishing -> finishing;
        issue-reported -> issue;
        finished -> finished;
    }
    paused {
       stop-pause -> in_use;
    }
    finishing {
       finished -> finished;
    }
    released {}
    finished {}
    issue {}
    expired {}
    cancelled {}
  }
}

**Comment by edwinvandenbelt** [comment]
with 'asset-claimed':
![image](https://user-images.githubusercontent.com/10400054/64861538-7d45a000-d630-11e9-9906-d9f802869a50.png)

// UML state diagram for a Booking, represented in Umple

class Booking {
  state {
    availibility_checked { 
      availability-request -> pending;
    }
    pending {
        asset-request -> confirmed;
        cancel-request -> released;
        expired -> expired;
    }
    confirmed {
        cancel-booking -> cancelled;
        in-preparation -> preparing;
        in-use-request -> in_use;
        asset-claimed -> confirmed;
    }
    preparing {
       in-use-confirm -> in_use;
    }
    in_use {
        start-pause -> paused;
        start-finishing -> finishing;
        issue-reported -> issue;
        finished -> finished;
    }
    paused {
       stop-pause -> in_use;
    }
    finishing {
       finished -> finished;
    }
    released {}
    finished {}
    issue {}
    expired {}
    cancelled {}
  }
}

**Comment by edwinvandenbelt** [comment]
With these STD's you can easily see the states and the allowed transitions. 
I've found some points to discuss:
1) is it necessary to distinguish '_released_' and '_expired_'?
2) is there only 1 exit state from _preparing_ (I don't think so)
3) same for _paused_
4) same for _finishing_
5) is _issue_ an exit state?

**Comment by edwinvandenbelt** [comment]
I think this one is pretty complete:
![image](https://user-images.githubusercontent.com/10400054/64863545-5dfd4180-d635-11e9-9b75-8af837e16e53.png)
// UML state diagram for a Booking, represented in Umple

class Booking {
  state {
    planning_phase {
        availability-request -> booking_phase;
    }
    booking_phase {
      pending {
          asset-request -> confirmed;
          cancel-request -> released;
          expired -> expired;
      }
      confirmed {
          cancel-booking -> cancelled;
          in-preparation -> preparing;
          in-use-request -> in_use;
          asset-claimed -> asset_claimed;
      }
      released {}
      expired {}
    }
    trip_execution_phase {
      preparing {
         in-use-confirm -> in_use;
         cancelled -> cancelled;
      }
      asset_claimed {
          in-preparation -> preparing;
          in-use-request -> in_use;
          asset-claimed -> asset_claimed;
          cancelled -> cancelled;
      }
      in_use {
          start-pause -> paused;
          start-finishing -> finishing;
          finished -> finished;
      }
      paused {
         stop-pause -> in_use;
         finished -> finished;
         start-finishing -> finishing;
      }
      finishing {
         finished -> finished;
      }
      finished {}
      issue-reported -> issue;
      
    }
    issue {}
    cancelled {}
  }
}

**Comment by edwinvandenbelt** [comment]
![image](https://user-images.githubusercontent.com/10400054/65247421-a9f62d80-daf0-11e9-966d-9a78add67a65.png)

**Comment by edwinvandenbelt** [comment]
Happy flow:
![image](https://user-images.githubusercontent.com/10400054/65248999-6ea92e00-daf3-11e9-960f-135d8607f021.png)

app->+MSP : find options
note over MSP: find routes & TO's on route
MSP->+TO1 : availablity-check
MSP->+TO2 : availablity-check
MSP->+TO3 : availablity-check
TO3->-MSP : no legs
TO2->-MSP : some car legs
TO1->-MSP : some bike legs
note over MSP: construct trip using the legs
MSP->+TO1: availibility-check-with-ids
TO1->-MSP: [1,34,52]
MSP->+TO2: availibility-check-with-ids 
TO2->-MSP: [3]
MSP->-app: propose trips
app->+MSP: select trip: 2 legs: car and bike

MSP->+TO1: availability-request [34]
MSP->+TO2: availability-request [3]
TO1->-MSP: ack [34]
TO2->-MSP: ack [3]

MSP->+TO1: asset-request [34]
MSP->+TO2: asset-request [3]
TO1->-MSP: ack [34]
TO2->-MSP: ack [3]

MSP->app: booking confirmed

**Comment by edwinvandenbelt** [comment]
Example not happy flow:
![image](https://user-images.githubusercontent.com/10400054/65249520-68678180-daf4-11e9-9ba9-f4b3738369c2.png)

**Comment by edwinvandenbelt** [comment]
Extra comment: the availability-request is to 'simulate' a transaction. After the availability-request the asset-request is a 'commit': every leg/asset is 'claimed'. After every TO has acknowledged the leg/asset booking, the asset-request is not allowed to respond a 'not acknowledged'.

**Comment by pimmeh** [comment]
Is the current set of events, states, etcetera adequately specified for version 1.1? I feel we are into the details now and should finish it up for now.

**Comment by JRobertoReyesG** [comment]
I agree. The last updates proposed by Edwin (availability check & availability checks with ids) are included in the process flow diagrams. I think we can close this issue now
